Source/Urho3D/CSharp/Managed/Container/SharedPtr.cs

```diff
@@ -0,0 +1,146 @@
+//
+// Copyright (c) 2022-2022 the rbfx project.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+
+namespace Urho3DNet
+{
+    public class SharedPtr<T> : IDisposable, IEquatable<SharedPtr<T>> where T : RefCounted
+    {
+        /// <summary>
+        /// Construct immutable SharedPtr object.
+        /// Don't forget to dispose it.
+        /// </summary>
+        /// <param name="other">Object to store pointer to.</param>
+        public SharedPtr(T other)
+        {
+            if (other?.IsNotExpired == true)
+            {
+                _value = other;
+                _value.AddRef();
+            }
+            else
+                _value = null;
+        }
+
+        /// <summary>
+        /// SharedPtr finalizer.
+        /// Should not be executed on not-disposed SharedPtr.
+        /// </summary>
+        ~SharedPtr()
+        {
+            if (_value?.IsNotExpired == true)
```

Comment:
This hides a potential bug too. Say we call `Dispose()`, currently it does not set `_value` to `null` so null, also say there are multiple strong references (a second `SharedPtr<>` on either C++ or C# side), then `IsNotExpired == true` even though we did call `Dispose()` on this instance. If we do set `_value` to `null` in `Dispose()` like i suggested, then this still should be changed to `_value != null` to be more explicit and not imply that `_value` can be non-null and there still may be no problem with it (we didnt call `Dispose()` on this object, but we errorneously called extra `ReleaseRef()` elsewhere, making everything work by chance, while application behavior is actually not correct).